### PR TITLE
switch to yarn, live a happy life

### LINKED ISSUE #1806: update to npm@5 for great justice
Opening up this issue to track upgrading all the things and do some testing:

 - [ ] local testing of new `package-lock.json`
 - [ ] CI builds
 - [ ] documentation
 - [ ] giving everyone time to get up to date before merging cc @desktop/core

Random pointers from the [release notes](https://github.com/npm/npm/releases/tag/v5.0.0):

 - `npm` will `--save` by default now.
 - better use of the local cache, for situations where you have no internet access
 - [`npm install` is sooooo much faster](https://twitter.com/maybekatz/status/865393382260056064)

TODO:

 - will we get a new version of node-gyp which fixes #1766 ?

### PR DESCRIPTION
**Pull request by shiftkey**
Fixes #1806

TL;DR: i keep wanting to catch up to the latest version of NPM but I keep encountering platform-specific issues despite it being four months since the first v5 release. Yarn however doesn't seem to have these issues. So why not switch?

 - [x] CI platforms all pass
 - [x] update docs
 - [x] lock file is stable between platforms - ~~https://github.com/npm/npm/issues/17722~~ let's do yarn
 - [x] cleanup branch to be less caremad
 - [x] write some docs around updating packages
 - [x] merge in a few updates and ensure it's all fine
 - [x] do a deploy to the test channel and verify the app works
   - [x] macOS
   - [x] Windows
 - [x] yarn 1.2 just landed, does that help?
 - [ ] give contributors time to update their local machines (2 weeks?)

**TODO:**

For anyone working with the Desktop repo:

```
> npm install -g yarn@1.2.0
```

**Documentation to read:**

 - [Updated Setup Guide](https://github.com/desktop/desktop/blob/74331f62e538aa5bd4fa999a5543425afc7ab7f5/docs/contributing/setup.md)
 - [Cheat Sheet for Yarn](https://github.com/desktop/desktop/blob/00ad4153f4d3f8ead878499cf694ae4696015752/docs/contributing/working-with-packages.md)
- [`npm` commands and their `yarn` equivalents](https://yarnpkg.com/en/docs/migrating-from-npm)

### CODE REVIEW

**Comment by shiftkey** [comment]
lol, while I committed the new `package-lock.json` file on Windows I then ran an `npm i` on macOS and it chooses a different integrity algorithm for the packages

```diff
     "codemirror": {
       "version": "5.29.0",
       "resolved": "https://registry.npmjs.org/codemirror/-/codemirror-5.29.0.tgz",
-      "integrity": "sha512-nlG9m0YQ0gFhdEdnKDG+XJRB/bW+K6M9Axs01+LScjVamWtd4dEwgyohf/r4voW1efnGi6U6hHHvDQ9tt9BtoA=="
+      "integrity": "sha1-5o3hNQ4vDOgEo5MFdtCuMYc26Wc="
     },
```

**Comment by shiftkey** [comment]
Now the `fsevents` package is giving me grief. I'm not sure what the root dependendency is, it's required by `chokidar` which is a wrapper around filesystem events.

It's installed for macOS and not for Windows. This is annoying because an `npm install` will edit the `package-lock.json` file to add or remove the file on rival OSes. This is https://github.com/npm/npm/issues/17722, and there's some workarounds discussed in that that might work for us.

**Comment by shiftkey** [comment]
Alright, looks like it's time to try yarn and see if it can do better...

**Comment by j-f1** [comment]
Looks like this also fixes #2800 🎉

I’ll revert it when this PR gets merged.

**Comment by shiftkey** [comment]
@desktop/core this is getting close to being ready to review. Would love some feels on this and what I can do to help make this transition smooth when you have a moment...

File: docs/contributing/setup.md [hunk 1]
```diff
@@ -49,7 +58,7 @@ repository.
 After cloning the repository, the typical workflow to get up running
 is as follows:
 
-* Run `npm install` to get all required dependencies on your machine.
+* Run `yarn` to get all required dependencies on your machine.
 * Run `npm run build:dev` to create a development build of the app.
```
Comment (joshaber) [inline docs/contributing/setup.md]: Is there some `yarn` equivalent to `npm run`? Feels a little weird to use a mix of both, but maybe that's ok.

File: appveyor.yml [hunk 2]
```diff
@@ -21,13 +21,9 @@ version: "{build}"
 
 install:
   - ps: Install-Product node $env:nodejs_version $env:platform
-  - npm install -g npm@4.6.1
+  - npm install -g yarn@1.1.0
   - git submodule update --init --recursive
-  - npm prune
```
Comment (joshaber) [inline appveyor.yml]: Why don't we prune anymore?

File: docs/contributing/setup.md (hunk 1 above)
Comment (j-f1) [reply docs/contributing/setup.md]: `yarn foo` = `npm run foo`

File: appveyor.yml (hunk 2 above)
Comment (j-f1) [reply appveyor.yml]: I *think* `yarn` does it automatically.

Comment (shiftkey) [reply appveyor.yml]:
```sh
$ yarn prune
yarn prune v1.1.0
...
error The prune command isn't necessary. `yarn install` will prune extraneous packages.
info Visit https://yarnpkg.com/en/docs/cli/prune for documentation about this command.
```

File: docs/contributing/setup.md (hunk 1 above)
Comment (shiftkey) [reply docs/contributing/setup.md]: Oh snap, thanks for the hint @j-f1! That's gonna save some keystrokes! Will tidy up.

File: circle.yml [hunk 3]
```diff
@@ -20,21 +20,21 @@ dependencies:
 
   override:
     - yarn
-    - npm rebuild
+    - yarn rebuild
```
Comment (j-f1) [inline circle.yml]: The yarn docs [say](https://yarnpkg.com/en/docs/migrating-from-npm) that this should be `yarn install --force`.

File: circle.yml [hunk 4]
```diff
@@ -20,21 +20,21 @@ dependencies:
...
-    - npm test
+    - yarn test:setup
+    - yarn test
 
 deployment:
   production:
     branch: /^__release-.*/
     commands:
-      - npm run publish
+      - yarn publish
```
Comment (j-f1) [inline circle.yml]: This should be `yarn run publish` — `yarn publish` is like `npm publish`.

**Comment by shiftkey** [comment]
A quick comparison of the CI times on current `master` and this PR:

|          | [before](https://github.com/desktop/desktop/commit/1b04319b039cbab482d3505acd53fd6a6bf434ae) | [after](https://github.com/desktop/desktop/commit/3cbbdab51af94c997dbf7402353b7943115b1139) |
|----------|----------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------|
| Circle        | 11min, 8sec             | 9min, 51sec      |
| AppVeyor | 11min, 8sec .           | 8min, 28sec      |
| Travis       | 10min, 25sec           | 11min, 30 sec*  |

* The previous 3 Travis build times were 9min 33sec, 9min 28sec and 9min 50sec. LOL Travis.

**Comment by joshaber** [comment]
Let's plan to merge this early next week 🎉

**Comment by joshaber** [comment]
giphy-2
